lgc/util/ElfNoteEntryInsertionUtil.cpp

```diff
@@ -0,0 +1,258 @@
+
+  // Write the new contents.
+  raw_svector_ostream elfStream(elf);
+  ArrayRef<uint8_t> noteEntry;
+  auto error = elfContentStream.readBytes(0, elfContentStream.getLength(), noteEntry);
+  (void)error;
+  assert(!error);
+  elfStream << StringRef(reinterpret_cast<const char *>(noteEntry.data()), noteEntry.size());
+
+  // Write the sections after the insertion offset.
+  for (const auto &sectionAndNewOffsetInfo : sectionAndNewOffset) {
+    elfStream.write_zeros(sectionAndNewOffsetInfo.newOffset - elfStream.str().size());
+    elfStream << sectionAndNewOffsetInfo.section.str();
+  }
+}
+
+// =====================================================================================================================
+// Write the section header table to the given offset.
+//
+// @param [in/out] elf             : The ELF to write the section header table.
+// @param sectionHeaderTableOffset : The offset where it will write the section header table.
+// @param sectionHeaderTable       : The section header table to be written to the ELF.
+void writeSectionHeaderTable(SmallVectorImpl<char> &elf, const ELF::Elf64_Off sectionHeaderTableOffset,
+                             const SmallVectorImpl<ELF::Elf64_Shdr> &sectionHeaderTable) {
+  if (sectionHeaderTable.size() == 0)
+    return;
+
+  raw_svector_ostream elfStream(elf);
+  unsigned minElfSizeForSectionHeaders = sectionHeaderTableOffset + sizeof(ELF::Elf64_Shdr) * sectionHeaderTable.size();
+  if (minElfSizeForSectionHeaders > elf.size())
+    elfStream.write_zeros(minElfSizeForSectionHeaders - elf.size());
+  elfStream.pwrite(reinterpret_cast<const char *>(sectionHeaderTable.data()),
+                   sizeof(ELF::Elf64_Shdr) * sectionHeaderTable.size(), sectionHeaderTableOffset);
+}
+
+} // anonymous namespace
+
+namespace lgc {
+
+// =====================================================================================================================
+// Adds the given note entries to the given ELF if the given ELF has a note section.
+// Otherwise, it does nothing.
+//
+// @param [in/out] elf : ELF to be updated with the new note entries.
+// @param notes        : An array of note entries to be inserted to the existing note section.
+void AddNotesToELF(SmallVectorImpl<char> &elf, llvm::ArrayRef<NoteEntry> notes) {
+  // Get ELF header that contains information for section header table offset
+  // and the number of section headers.
+  //
+  // Reference: http://www.skyfree.org/linux/references/ELF_Format.pdf
+  ELF::Elf64_Ehdr *ehdr = reinterpret_cast<ELF::Elf64_Ehdr *>(elf.data());
+
+  // Get the section headers and the existing note section.
+  MutableArrayRef<ELF::Elf64_Shdr> sectionHeaders(reinterpret_cast<ELF::Elf64_Shdr *>(elf.data() + ehdr->e_shoff),
+                                                  ehdr->e_shnum);
+  auto existingNoteSection = find_if(
+      sectionHeaders, [](const ELF::Elf64_Shdr &sectionHeader) { return sectionHeader.sh_type == ELF::SHT_NOTE; });
```

Comment:
Do you want to add it to any note section or to a specific one named ".note"?

Reply:
It might be worth making the section name a parameter to this function to make it more general.

Reply:
done